handshake_server.go

```diff
@@ -299,11 +304,48 @@ Curves:
 		c.sendAlert(alertInternalError)
 		return false, err
 	}
-	if hs.clientHello.scts && hs.hello != nil {
+
+	// Set the private key for this handshake to the certificate's secret key.
+	hs.privateKey = hs.cert.PrivateKey
+
+	if hs.clientHello.scts {
 		hs.hello.scts = hs.cert.SignedCertificateTimestamps
 	}
 
-	if priv, ok := hs.cert.PrivateKey.(crypto.Signer); ok {
+	// Set the private key to the DC private key if the client and server are
+	// willing to negotiate the delegated credential extension.
+	//
+	// Check to see if a DelegatedCredential is available and should be used.
+	// If one is available, the session is using TLS >= 1.2, and the client
+	// accepts the delegated credential extension, then set the handshake
+	// private key to the DC private key.
+	if c.config.GetDelegatedCredential != nil && hs.clientHello.delegatedCredential && c.vers >= VersionTLS12 {
```

Comment:
Another quick question - "GetConfigForClient" this can't be reused here?

Reply:
This could be an option, but again, the TLS version chosen by the server can't be determined from `ClientHelloInfo` (the input to `GetConfigForClient`). This is needed because the delegation binds the DC to the protocol.

One way we could get around this is by having an internal `map[uint16]*DelegatedCredential` field of the serverHandshakeState structure. But we'd have to expose an interface for populating this. I thought it would be better to leave how DCs are chosen to the deployment, hence the `GetDelegatedCredential` callback. This also aligns nicely with `GetCertificate`.